extensions/amp-iframely/0.1/amp-iframely.js

```diff
@@ -0,0 +1,417 @@
+      TAG,
+      this.element
+    );
+    if (!this.widgetId_) {
+      if (this.url_) {
+        userAssert(
+          this.key_,
+          'Iframely data-key must also be set when you specify data-url parameter at <%s> %s',
+          TAG,
+          this.element
+        );
+      }
+      if (this.key_) {
+        userAssert(
+          this.url_,
+          'Iframely data-url must also be set when you specify data-key parameter at <%s> %s',
+          TAG,
+          this.element
+        );
+        userAssert(
+          16 < this.key_.length || this.key_.length > 256,
+          'Iframely data-key should be between 16 and 256 characters parameter at <%s> %s',
+          TAG,
+          this.element
+        );
+      }
+      if (this.key_ || this.url_) {
+        userAssert(
+          !this.widgetId_,
+          'Iframely data-id should not be set when there is already a pair of data-url and data-key for <%s> %s',
+          TAG,
+          this.element
+        );
+      }
+    }
+    if ((this.widgetId_ && this.url_) || (this.widgetId_ && this.key_)) {
+      userAssert(
+        !this.widgetId_,
+        'Only one way of setting either data-id or data-url and data-key supported for <%s> %s',
+        TAG,
+        this.element
+      );
+    }
+    if (this.widgetId_) {
+      this.src_ = addParamToUrl(this.base_ + this.widgetId_, 'amp', '1');
+    } else {
+      this.src_ = addParamsToUrl(this.base_ + 'api/iframe', {
+        'url': this.url_,
+        'key': this.key_,
+        'amp': '1',
+      });
+    }
+  }
+
+  /**
+   * Parse other data-* attributes and append them to API query url
+   * @return {object} of Iframely options
+   * @private
+   * */
+  parseOptions_() {
+    const options = {};
+    const exclude = [
+      'data-id',
+      'data-domain',
+      'data-key',
+      'data-url',
+      'data-img',
+    ];
+    let data = this.element
+      .getAttributeNames()
+      .filter(name => name.startsWith('data-'));
+    data = data.filter(name => !exclude.includes(name));
+    data.forEach(
+      item =>
+        (options[item.split('data-').pop()] = this.element.getAttribute(item))
+    );
+    return options;
+  }
+
+  /** @override */
+  unlayoutCallback() {
+    if (this.iframe_) {
+      removeElement(this.iframe_);
+      this.iframe_ = null;
+    }
+    return true; // Request layoutCallback again.
+  }
+
+  /**
+   * Validates that requested domain is a valid Iframely domain
+   * @param {string} domainName - of domain to check against a whitelist
+   * @return {boolean} if domain is valid
+   * @private
+   * */
+  isValidDomain_(domainName) {
+    const whitelistedDomains = [
+      /^(?:[^\.\/]+\.)?iframe\.ly$/i,
+      /^if\-cdn\.com$/i,
+      /^iframely\.net$/i,
+      /^oembed\.vice\.com$/i,
+      /^iframe\.nbcnews\.com$/i,
+    ];
+    for (const i in whitelistedDomains) {
+      if (whitelistedDomains[i].test(domainName)) {
+        return true;
+      }
+    }
+    return false;
```

Comment:
1. We prefer "allowlist". "Allowed" may be more appropriate as an adjective here.
2. [`for (x in array)` is bad.](https://stackoverflow.com/questions/500504/why-is-using-for-in-for-array-iteration-a-bad-idea)

```suggestion
    const allowedDomains = [
      /^(?:[^\.\/]+\.)?iframe\.ly$/i,
      /^if\-cdn\.com$/i,
      /^iframely\.net$/i,
      /^oembed\.vice\.com$/i,
      /^iframe\.nbcnews\.com$/i,
    ];
    return allowedDomains.reduce(
      (allowed, re) => allowed || re.test(domainName),
      false
    );
```